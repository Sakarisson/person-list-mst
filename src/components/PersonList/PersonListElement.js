import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import styled from 'styled-components';

const Container = styled.div``;

const PersonListElement = ({ person: { id, fullName } }) => (
  <Container>
    <Link to={`/people/${id}`}>{fullName}</Link>
  </Container>
);

PersonListElement.propTypes = {
  person: PropTypes.shape({
    fullName: PropTypes.string.isRequired,
  }).isRequired,
};

export default PersonListElement;
