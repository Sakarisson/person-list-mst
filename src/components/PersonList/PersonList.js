import React from 'react';
import PropTypes from 'prop-types';
import { observer } from 'mobx-react';
import { compose } from 'recompose';

import withStore from '../../hoc/withStore';
import PersonListElement from './PersonListElement';

const PersonList = ({ store }) => (
  <div>
    {store.people.map(person => (
      <PersonListElement
        person={person}
        remove={() => store.removePerson(person.id)}
        key={person.id}
      />
    ))}
  </div>
);

PersonList.propTypes = {
  store: PropTypes.shape({
    people: PropTypes.arrayOf(PropTypes.object).isRequired,
  }).isRequired,
};

export default compose(
  withStore,
  observer,
)(PersonList);
