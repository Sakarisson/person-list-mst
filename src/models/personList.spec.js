import PersonList from './personList';

const mockPeople = [
  {
    id: '1',
    firstName: 'Bob',
    lastName: 'Allison',
    address: { streetAddress: 'a', zipCode: 'a', city: 'a' },
    friends: ['2'],
  },
  {
    id: '2',
    firstName: 'Charles',
    lastName: 'Bobson',
    address: { streetAddress: 'a', zipCode: 'a', city: 'a' },
    friends: ['1', '3'],
  },
  {
    id: '3',
    firstName: 'Alice',
    lastName: 'Charlson',
    address: { streetAddress: 'a', zipCode: 'a', city: 'a' },
    friends: ['2'],
  },
];

describe('personList', () => {
  describe('model', () => {
    it('initializes properly', () => {
      expect(PersonList.create()).toMatchSnapshot();
    });

    it('initializes with existing people', () => {
      expect(
        PersonList.create({
          people: mockPeople,
        }),
      ).toMatchSnapshot();
    });

    it('initializes with sortBy firstName', () => {
      expect(PersonList.create({ sortBy: 'firstName' })).toMatchObject({ sortBy: 'firstName' });
    });

    it('initializes with sortBy lastName', () => {
      expect(PersonList.create({ sortBy: 'lastName' })).toMatchObject({ sortBy: 'lastName' });
    });

    it('initializes with sortOrder ascending', () => {
      expect(PersonList.create({ sortOrder: 'ascending' })).toMatchObject({
        sortOrder: 'ascending',
      });
    });

    it('initializes with sortOrder descending', () => {
      expect(PersonList.create({ sortOrder: 'descending' })).toMatchObject({
        sortOrder: 'descending',
      });
    });
  });

  describe('actions', () => {
    it('adds person', () => {
      const personList = PersonList.create({
        people: [
          {
            id: '1',
            firstName: 'Person',
            lastName: 'Persson',
            address: { streetAddress: 'Address', zipCode: '123', city: 'City' },
          },
        ],
      });
      personList.addPerson({
        id: '2',
        firstName: 'Jane',
        lastName: 'Doe',
        address: { streetAddress: 'Weirdplace', zipCode: '321', city: 'Town' },
      });
      expect(personList).toMatchSnapshot();
    });

    it('removes person', () => {
      const personList = PersonList.create({
        people: mockPeople,
      });
      personList.removePerson('1');
      expect(personList.people.map(p => p.id)).toEqual(['2', '3']);
    });

    it('clears all friends', () => {
      const personList = PersonList.create({ people: mockPeople });
      personList.clearAllFriends();
      expect(personList.people.map(p => p.friends)).toEqual([[], [], []]);
    });

    it('sets sortBy to none, firstName and lastName', () => {
      const personList = PersonList.create({});
      personList.setSortBy('none');
      expect(personList.sortBy).toBe('none');
      personList.setSortBy('firstName');
      expect(personList.sortBy).toBe('firstName');
      personList.setSortBy('lastName');
      expect(personList.sortBy).toBe('lastName');
    });

    it('sets sortOrder to ascending and descending', () => {
      const personList = PersonList.create({});
      personList.setSortOrder('ascending');
      expect(personList.sortOrder).toBe('ascending');
      personList.setSortOrder('descending');
      expect(personList.sortOrder).toBe('descending');
    });
  });

  describe('views', () => {
    it('sorts by firstName in ascending order', () => {
      const personList = PersonList.create({
        sortBy: 'firstName',
        sortOrder: 'ascending',
        people: mockPeople,
      });
      expect(personList.sortedPeople.map(p => p.id)).toEqual(['3', '1', '2']);
    });

    it('sorts by firstName in descending order', () => {
      const personList = PersonList.create({
        sortBy: 'firstName',
        sortOrder: 'descending',
        people: mockPeople,
      });
      expect(personList.sortedPeople.map(p => p.id)).toEqual(['2', '1', '3']);
    });

    it('sorts by lastName in ascending order', () => {
      const personList = PersonList.create({
        sortBy: 'lastName',
        sortOrder: 'ascending',
        people: mockPeople,
      });
      expect(personList.sortedPeople.map(p => p.id)).toEqual(['1', '2', '3']);
    });

    it('sorts by lastName in descending order', () => {
      const personList = PersonList.create({
        sortBy: 'lastName',
        sortOrder: 'descending',
        people: mockPeople,
      });
      expect(personList.sortedPeople.map(p => p.id)).toEqual(['3', '2', '1']);
    });
  });
});
