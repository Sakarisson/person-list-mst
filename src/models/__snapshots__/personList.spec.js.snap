// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`personList actions adds person 1`] = `
Object {
  "people": Array [
    Object {
      "address": Object {
        "city": "City",
        "streetAddress": "Address",
        "zipCode": "123",
      },
      "firstName": "Person",
      "friends": Array [],
      "friendsSortBy": "none",
      "friendsSortOrder": "ascending",
      "id": "1",
      "lastName": "Persson",
    },
    Object {
      "address": Object {
        "city": "Town",
        "streetAddress": "Weirdplace",
        "zipCode": "321",
      },
      "firstName": "Jane",
      "friends": Array [],
      "friendsSortBy": "none",
      "friendsSortOrder": "ascending",
      "id": "2",
      "lastName": "Doe",
    },
  ],
  "sortBy": "none",
  "sortOrder": "ascending",
}
`;

exports[`personList model initializes properly 1`] = `
Object {
  "people": Array [],
  "sortBy": "none",
  "sortOrder": "ascending",
}
`;

exports[`personList model initializes with existing people 1`] = `
Object {
  "people": Array [
    Object {
      "address": Object {
        "city": "a",
        "streetAddress": "a",
        "zipCode": "a",
      },
      "firstName": "Bob",
      "friends": Array [
        "2",
      ],
      "friendsSortBy": "none",
      "friendsSortOrder": "ascending",
      "id": "1",
      "lastName": "Allison",
    },
    Object {
      "address": Object {
        "city": "a",
        "streetAddress": "a",
        "zipCode": "a",
      },
      "firstName": "Charles",
      "friends": Array [
        "1",
        "3",
      ],
      "friendsSortBy": "none",
      "friendsSortOrder": "ascending",
      "id": "2",
      "lastName": "Bobson",
    },
    Object {
      "address": Object {
        "city": "a",
        "streetAddress": "a",
        "zipCode": "a",
      },
      "firstName": "Alice",
      "friends": Array [
        "2",
      ],
      "friendsSortBy": "none",
      "friendsSortOrder": "ascending",
      "id": "3",
      "lastName": "Charlson",
    },
  ],
  "sortBy": "none",
  "sortOrder": "ascending",
}
`;
